lxd/storage/drivers/driver_pure_util.go

```diff
@@ -517,6 +572,475 @@ func (p *pureClient) disconnectHostFromVolume(poolName string, volName string, h
+			if shared.PathExists(path) {
+				err := os.WriteFile(path, []byte("1"), 0400)
+				if err != nil {
+					return fmt.Errorf("Failed to unmap volume %q: Failed to remove device %q: %w", vol.name, devName, err)
+				}
+			}
+		}
+
+		ctx, cancel := context.WithTimeout(d.state.ShutdownCtx, 10*time.Second)
+		defer cancel()
+
+		if !waitGone(ctx, volumePath) {
+			return fmt.Errorf("Timeout exceeded waiting for PureStorage volume %q to disappear on path %q", vol.name, volumePath)
+		}
+	}
+
+	// If this was the last volume being unmapped from this system, terminate iSCSI session
+	// and remove the host from PureStorage.
+	if host.ConnectionCount == 1 {
+		target, err := d.client().getTarget()
+		if err != nil {
+			return err
+		}
+
+		// Disconnect from the target.
+		err = d.disconnect(*target)
+		if err != nil {
+			return err
+		}
+
+		// Remove the host from PureStorage.
+		err = d.client().deleteHost(host.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getMappedDevPath returns the local device path for the given volume.
+// Indicate with mapVolume if the volume should get mapped to the system if it isn't present.
+func (d *pure) getMappedDevPath(vol Volume, mapVolume bool) (string, revert.Hook, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	if mapVolume {
+		err := d.mapVolume(vol)
+		if err != nil {
+			return "", nil, err
+		}
+
+		revert.Add(func() { _ = d.unmapVolume(vol) })
+	}
+
+	// findDevPathFunc has to be called in a loop with a set timeout to ensure
+	// all the necessary directories and devices can be discovered.
+	findDevPathFunc := func(diskPrefix string, diskSuffix string) (string, error) {
+		var diskPaths []string
+
+		// If there are no other disks on the system by id, the directory might not even be there.
+		// Returns ENOENT in case the by-id/ directory does not exist.
+		diskPaths, err := resources.GetDisksByID(diskPrefix)
+		if err != nil {
+			return "", err
+		}
+
+		for _, diskPath := range diskPaths {
+			// Skip the disk if it is only a partition of the actual volume.
+			if strings.Contains(diskPath, "-part") {
+				continue
+			}
+
+			// Skip volumes that do not have volume's pool and name suffix.
+			if !strings.HasSuffix(diskPath, diskSuffix) {
+				continue
+			}
+
+			// The actual device might not already be created.
+			// Returns ENOENT in case the device does not exist.
+			devPath, err := filepath.EvalSymlinks(diskPath)
+			if err != nil {
+				return "", err
+			}
+
+			return devPath, nil
+		}
+
+		return "", nil
+	}
+
+	var volumeDevPath string
+	var diskPrefix string
+	var diskSuffix string
+
+	volName, err := d.getVolumeName(vol)
+	if err != nil {
+		return "", nil, err
+	}
+
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		diskPrefix = "scsi-"
+		diskSuffix = fmt.Sprintf("%s::%s", vol.pool, volName)
+	}
+
+	timeout := time.Now().Add(15 * time.Second)
+
+	// It might take a while to create the local disk.
+	// Retry until it can be found.
+	for {
```

Comment:
Please can we use a pattern like `tryExists` or `waitGone` that accepts a context with a WithTimeout, that way we can include the state's shutdown context too and immediately return on next iteration rather than keeping going until the timeout hits.